### ISSUE TITLE
Feature request: add check.array.of(obj, pred) predicate

### ISSUE THREAD

**Issue by ceineke** [philbooth/check-types.js#28]
Add a chainable predicate named "of" to that checks that an the object under scrutiny is
a.) an array
b.) contains only elements that pass the predicate.

For example:
check.verify.array.of.string(arr);
check.verify.array.of(arr, function (el) { return check.instance(el, Foo) || check.instance(el, Bar); });

**Comment by MareoRaft** [comment]
I am a big fan of ceineke's suggestion.  Something like

    check.array.of.number(input)

would be super nice.  I think I ended up using something very ugly like `check.all(check.apply(array,check.number))`.  There's also the idea that `.all` could take in any iterable, giving a nice generic use like

    check.all.number(input)

Thanks for your great library and thanks for considering this feature!

**Comment by philbooth** [comment]
Ha, okay, I shall re-open this one in that case. :)

**Comment by ceineke** [comment]
I like the idea of making ```all``` check any iterable, although that mean that it precedes a check.array or check.object. So, check.verify.all.number.or.string would throw Error if the given object isn't an array of numbers and strings, or an object where the values are numbers and strings. Can we get all three?

```
check.all.number.or.unemptyString
check.array.of.number.or.unemptyString
check.object.of.number.or.unemptyString
```
?

**Comment by MareoRaft** [comment]
All three would be great.  I think it is a mistake to think that an iterable must be an array or an object.  Iterables are very general.  They are basically anything which provides a special `next()` function that allows you to iterate through all the values.  For example, it could be a linked list or a class which you define yourself.  That is why an `all` option could be very powerful.  Of course, the programmer can pick which tool to use for their specific needs.

**Comment by philbooth** [comment]
It's unclear in that code snippet which bits you think should be functions. Where do the parentheses go and what are the arguments in each case?

I've pushed a branch containing unit tests for one possible implementation of `of`:

https://github.com/philbooth/check-types.js/compare/of?expand=1

Is that interface good or bad? How would you change it?

**Comment by MareoRaft** [comment]
Here are some use cases:

`check.array.of.even(A)` would first check that A is an array.  If it succeeds, it then checks that all elements of A are even.

Everything I see is perfect with the following exceptions / clarifications:

1.

`check.object.of.either.odd.or.even(B)`  Well here is something interesting.  I would want this to first check that B is an object, then if it succeeds, check that each element of B is either `odd` or `even`.  I see that your test on line `1213` says that we can't apply `either`.  If we can apply something to a single value, then we should be able to apply it to an array too.  But I think I've changed the usage here.  This opens up a whole new can of worms about how we expect `or` to behave.  If you are interested in discussing it, create a separate thread and message me.  Otherwise we can just table this.

2.

I see that you have chosen to use a single `of` for everything.  I think it was smart of you to recognize that the `of` and `all` above were really the same thing.  It's just that in the English language, it made sense to use `all` for stuff like `check.all.integer(A)` because it sounds like "check that all elements are integers"; and it made sense to use `of` for stuff like `check.array.of.integer(A)` because it sounds like "check that it's an array of integers".  I didn't realize that at first.

I would like to propose a compromise between a consistent computer grammar and an intuitive English grammar.  This would be to make the functions `of`, `all`, `each`, and `every` synonyms.  Everything can be implemented as `of`.  Then `all`, `each`, and `every` would merely be pointers to `of`.  That would allow the programmer to type what intuitively makes sense to them and know that it will work.

**Comment by philbooth** [comment]
Yeah, your point `1` is rather tricky to implement I think. At least initially, I'm going to focus on delivering something simpler that can serve as a foundation. It can then be incrementally improved with pull requests.

Regarding point `2`, I am a huge believer in Joshua Bloch's advice about APIs generally and the balance between completeness and minimalism specifically. It's worth finding one of his lectures online but I'll try to paraphrase him from memory: An interface should be complete enough to facilitate most use cases, but minimal so as to avoid confusion and promote consistency. I think having synonyms really fails the second part of that . A smaller interface is easier to learn, easier to remember, easier to predict and easier to reason about.

**Comment by MareoRaft** [comment]
In response to point 2, I think that is reasonable.  Moreover, ES6 is introducing a new `for (i of iterable)` loop that takes in any iterable and loops through the *values*.  Therefore, I think you are making the right decision to use `of` as your keyword.  (although if i ever write my own programming language, i will definitely be using *all*!)

**Comment by philbooth** [comment]
Just wanted to update you guys because I've finally got round to working on this properly. It's not finished yet but if you want to have a look at the approach I'm taking, it is in this branch:

https://github.com/philbooth/check-types.js/compare/issue_28?expand=1

If you have any comments or whatever, let me know.

**Comment by MareoRaft** [comment]
Glad to hear!  Did you decide to go with `arrayLike` instead of `iterable`?

I don't understand the code well enough to help you out.  But let me know when you are finished.  Looking forward to testing it out!

**Comment by philbooth** [comment]
Actually, I decided to go with both.

* `arrayLike` returns `true` for any object with a numeric `length` property.
* `iterable` returns `true` if the property `object[Symbol.iterator]` is a function, as per: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Iteration_protocols

I should have time to get it all wrapped up this week, will be sure to let you guys know.

**Comment by philbooth** [comment]
Fyi, this is now done in `master`. I just need to fix #41and document all the new hotness before I can publish on npm. Will post here when it is available. :)

**Comment by philbooth** [comment]
Published as 3.2.0, let me know if there's any issues. :)
